Maple2.File.Ingest/Mapper/AiMapper.cs

```diff
@@ -0,0 +1,204 @@
+﻿using Maple2.File.IO;
+using Maple2.File.Parser;
+using Maple2.File.Parser.Xml.AI;
+using Maple2.Model.Enum;
+using Maple2.Model.Metadata;
+
+namespace Maple2.File.Ingest.Mapper;
+
+public class AiMapper : TypeMapper<AiMetadata> {
+    private readonly AiParser parser;
+
+    public AiMapper(M2dReader xmlReader) {
+        parser = new AiParser(xmlReader);
+    }
+
+    protected override IEnumerable<AiMetadata> Map() {
+        foreach ((string name, NpcAi data) in parser.Parse()) {
+            List<AiMetadata.Condition> reserved = new List<AiMetadata.Condition>();
+            List<AiMetadata.Node> battle = new List<AiMetadata.Node>();
+            List<AiMetadata.Node> battleEnd = new List<AiMetadata.Node>();
+            List<AiMetadata.AiPresetDefinition> aiPresets = new List<AiMetadata.AiPresetDefinition>();
+
+            foreach (Condition node in data.reserved?.condition ?? new List<Condition>()) {
+                reserved.Add(MapCondition(node));
+            }
+
+            foreach (Node node in data.battle.node) {
+                battle.Add(MapNode(node));
+            }
+
+            foreach (Node node in data.battleEnd?.node ?? new List<Node>()) {
+                battleEnd.Add(MapNode(node));
+            }
+
+            foreach (AiPreset node in data.aiPresets?.aiPreset ?? new List<AiPreset>()) {
+                List<AiMetadata.Node> childNodes = new List<AiMetadata.Node>();
+                List<AiMetadata.AiPreset> childAiPresets = new List<AiMetadata.AiPreset>();
+
+                foreach (Node child in node.node) {
+                    childNodes.Add(MapNode(child));
+                }
+
+                foreach (AiPreset child in node.aiPreset) {
+                    childAiPresets.Add(new AiMetadata.AiPreset(child.name));
+                }
+
+                aiPresets.Add(new AiMetadata.AiPresetDefinition(node.name, childNodes.ToArray(), childAiPresets.ToArray()));
+            }
+
+            yield return new AiMetadata(name, reserved.ToArray(), battle.ToArray(), battleEnd.ToArray(), aiPresets.ToArray());
+        }
+    }
+
+    AiMetadata.Condition MapCondition(Condition node) {
+        List<AiMetadata.Node> childNodes = new List<AiMetadata.Node>();
+        List<AiMetadata.AiPreset> childAiPresets = new List<AiMetadata.AiPreset>();
+        
+        foreach (Node child in node.node) {
+            childNodes.Add(MapNode(child));
+        }
+
+        foreach (AiPreset child in node.aiPreset) {
+            childAiPresets.Add(new AiMetadata.AiPreset(child.name));
+        }
+
+        switch(node.name) {
+            case "distanceOver":
+                return new AiMetadata.DistanceOverCondition(node.name, childNodes.ToArray(), childAiPresets.ToArray(), node.value);
```

Comment:
For consistency, we intentionally name all parameters when parsing. This will ensure fields continue to map properly in the event that types change.

For example: https://github.com/AngeloTadeucci/Maple2/blob/master/Maple2.File.Ingest/Mapper/ItemMapper.cs#L157-L182